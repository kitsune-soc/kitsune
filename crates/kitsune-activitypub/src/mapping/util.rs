use super::State;
use kitsune_db::model::{account::Account, post::Visibility};
use kitsune_type::ap::PUBLIC_IDENTIFIER;

pub trait BaseToCc {
    fn base_to_cc(&self, state: State<'_>, account: &Account) -> (Vec<String>, Vec<String>);
}

impl BaseToCc for Visibility {
    #[inline]
    fn base_to_cc(&self, state: State<'_>, account: &Account) -> (Vec<String>, Vec<String>) {
        let followers_url = state.service.url.followers_url(account.id);

        match self {
            Visibility::Public => (vec![PUBLIC_IDENTIFIER.to_string()], vec![followers_url]),
            Visibility::Unlisted => (vec![], vec![PUBLIC_IDENTIFIER.to_string(), followers_url]),
            Visibility::FollowerOnly => (vec![followers_url], vec![]),
            Visibility::MentionOnly => (vec![], vec![]),
        }
    }
}
