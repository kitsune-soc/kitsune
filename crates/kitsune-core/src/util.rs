use crate::state::State;
use hex_simd::{AsOut, AsciiCase};
use iso8601_timestamp::Timestamp;
use kitsune_db::model::{account::Account, oauth2::access_token::AccessToken, post::Visibility};
use kitsune_type::ap::PUBLIC_IDENTIFIER;
use pulldown_cmark::{html, Options, Parser};

const TOKEN_LENGTH: usize = 32;

#[inline]
#[must_use]
pub fn generate_secret() -> String {
    let token_data: [u8; TOKEN_LENGTH] = rand::random();
    let mut buf = [0_u8; TOKEN_LENGTH * 2];
    (*hex_simd::encode_as_str(&token_data, buf.as_mut_slice().as_out(), AsciiCase::Lower))
        .to_string()
}

#[inline]
#[must_use]
pub fn process_markdown(markdown: &str) -> String {
    let parser = Parser::new_ext(markdown, Options::all());
    let mut buf = String::new();
    html::push_html(&mut buf, parser);
    buf
}

pub trait AccessTokenTtl {
    fn ttl(&self) -> time::Duration;
}

impl AccessTokenTtl for AccessToken {
    #[inline]
    fn ttl(&self) -> time::Duration {
        *self.expires_at - *Timestamp::now_utc()
    }
}

pub trait BaseToCc {
    fn base_to_cc(&self, state: &State, account: &Account) -> (Vec<String>, Vec<String>);
}

impl BaseToCc for Visibility {
    #[inline]
    fn base_to_cc(&self, state: &State, account: &Account) -> (Vec<String>, Vec<String>) {
        let followers_url = state.service.url.followers_url(account.id);

        match self {
            Visibility::Public => (vec![PUBLIC_IDENTIFIER.to_string()], vec![followers_url]),
            Visibility::Unlisted => (vec![], vec![PUBLIC_IDENTIFIER.to_string(), followers_url]),
            Visibility::FollowerOnly => (vec![followers_url], vec![]),
            Visibility::MentionOnly => (vec![], vec![]),
        }
    }
}
