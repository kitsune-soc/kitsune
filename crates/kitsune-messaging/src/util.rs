use std::fmt;

pub struct TransparentDebug<T>(pub T);

impl<T> fmt::Debug for TransparentDebug<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} {{ ... }}", std::any::type_name::<T>())
    }
}
