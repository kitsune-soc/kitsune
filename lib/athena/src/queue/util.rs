use ahash::AHashMap;
use redis::{streams::StreamId, FromRedisValue, RedisResult};

#[derive(Clone, Debug)]
pub struct StreamAutoClaimReply {
    pub start_stream_id: String,
    pub claimed_ids: Vec<StreamId>,
    pub deleted_ids: Vec<String>,
}

impl FromRedisValue for StreamAutoClaimReply {
    fn from_redis_value(v: &redis::Value) -> RedisResult<Self> {
        type AutoClaimReturnValue = (
            String,
            Vec<Vec<(String, AHashMap<String, redis::Value>)>>,
            Vec<String>,
        );

        let (start_stream_id, claimed_ids, deleted_ids): AutoClaimReturnValue =
            redis::from_redis_value(v)?;

        let claimed_ids: Vec<StreamId> = claimed_ids
            .into_iter()
            .flat_map(|row| row.into_iter().map(|(id, map)| StreamId { id, map }))
            .collect();

        Ok(Self {
            start_stream_id,
            claimed_ids,
            deleted_ids,
        })
    }
}
