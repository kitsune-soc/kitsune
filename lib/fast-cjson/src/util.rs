use sonic_rs::writer::WriteExt;
use std::{fmt, io, mem::MaybeUninit};

macro_rules! for_both {
    ($owner:ident, $matcher:pat => $impl:expr) => {{
        match $owner {
            Self::Left($matcher) => $impl,
            Self::Right($matcher) => $impl,
        }
    }};
}

pub enum Either<L, R> {
    Left(L),
    Right(R),
}

impl<L, R> io::Write for Either<L, R>
where
    L: io::Write,
    R: io::Write,
{
    #[inline]
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        for_both!(self, inner => inner.write(buf))
    }

    #[inline]
    fn write_all(&mut self, buf: &[u8]) -> io::Result<()> {
        for_both!(self, inner => inner.write_all(buf))
    }

    #[inline]
    fn write_fmt(&mut self, fmt: fmt::Arguments<'_>) -> io::Result<()> {
        for_both!(self, inner => inner.write_fmt(fmt))
    }

    #[inline]
    fn flush(&mut self) -> io::Result<()> {
        for_both!(self, inner => inner.flush())
    }
}

impl<L, R> WriteExt for Either<L, R>
where
    L: WriteExt + io::Write,
    R: WriteExt + io::Write,
{
    #[inline]
    fn reserve_with(&mut self, additional: usize) -> io::Result<&mut [MaybeUninit<u8>]> {
        for_both!(self, inner => inner.reserve_with(additional))
    }

    #[inline]
    #[allow(unsafe_code)] // We just dispatch over already unsafe implementations
    unsafe fn flush_len(&mut self, additional: usize) -> io::Result<()> {
        for_both!(self, inner => inner.flush_len(additional))
    }
}
