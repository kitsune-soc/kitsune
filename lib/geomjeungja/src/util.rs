use std::{fmt, ops::Deref};

#[derive(Clone)]
pub struct OpaqueDebug<T>(pub T);

impl<T> Deref for OpaqueDebug<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<T> fmt::Debug for OpaqueDebug<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct(std::any::type_name::<T>())
            .finish_non_exhaustive()
    }
}
