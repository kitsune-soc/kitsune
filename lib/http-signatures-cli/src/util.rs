use owo_colors::{OwoColorize, Stream};
use std::fmt::Display;

#[inline]
pub fn error_kaomoji() -> impl Display {
    "(┬┬﹏┬┬)".if_supports_color(Stream::Stdout, |text| text.red())
}

#[inline]
pub fn success_kaomoji() -> impl Display {
    "(^///^)".if_supports_color(Stream::Stdout, |text| text.green())
}
