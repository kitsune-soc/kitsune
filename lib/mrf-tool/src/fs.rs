use std::{
    collections::HashMap,
    fs::{self, File},
    io,
    ops::{Deref, DerefMut},
    path::{Path, PathBuf},
};

#[inline]
fn file_not_found() -> io::Error {
    io::Error::new(io::ErrorKind::NotFound, "file not found")
}

pub trait Filesystem {
    type File<'a>: io::Write
    where
        Self: 'a;

    fn copy(&mut self, src: &Path, dst: &Path) -> io::Result<()>;
    fn read(&mut self, path: &Path) -> io::Result<Vec<u8>>;

    fn create_or_truncate(&mut self, path: &Path) -> io::Result<Self::File<'_>>;
    fn open_append(&mut self, path: &Path) -> io::Result<Self::File<'_>>;
}

#[derive(Default)]
pub struct DummyFs {
    inner: HashMap<PathBuf, Vec<u8>>,
}

impl Deref for DummyFs {
    type Target = HashMap<PathBuf, Vec<u8>>;

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}

impl DerefMut for DummyFs {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.inner
    }
}

impl Filesystem for DummyFs {
    type File<'a> = &'a mut Vec<u8>;

    #[inline]
    fn copy(&mut self, src: &Path, dst: &Path) -> io::Result<()> {
        let value = self.inner.get(src).ok_or_else(file_not_found)?;
        self.inner.insert(dst.to_path_buf(), value.clone());

        Ok(())
    }

    #[inline]
    fn read(&mut self, path: &Path) -> io::Result<Vec<u8>> {
        self.inner.get(path).cloned().ok_or_else(file_not_found)
    }

    #[inline]
    fn create_or_truncate(&mut self, path: &Path) -> io::Result<Self::File<'_>> {
        Ok(self.inner.entry(path.to_path_buf()).or_default())
    }

    #[inline]
    fn open_append(&mut self, path: &Path) -> io::Result<Self::File<'_>> {
        self.inner.get_mut(path).ok_or_else(file_not_found)
    }
}

#[derive(Default)]
pub struct NativeFs {
    _priv: (),
}

impl Filesystem for NativeFs {
    type File<'a> = File;

    #[inline]
    fn copy(&mut self, src: &Path, dst: &Path) -> io::Result<()> {
        fs::copy(src, dst)?;
        Ok(())
    }

    #[inline]
    fn read(&mut self, path: &Path) -> io::Result<Vec<u8>> {
        fs::read(path)
    }

    #[inline]
    fn create_or_truncate(&mut self, path: &Path) -> io::Result<Self::File<'_>> {
        File::options()
            .create(true)
            .truncate(true)
            .write(true)
            .open(path)
    }

    #[inline]
    fn open_append(&mut self, path: &Path) -> io::Result<Self::File<'_>> {
        File::options().append(true).open(path)
    }
}
