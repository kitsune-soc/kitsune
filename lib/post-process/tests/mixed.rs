use crate::util::parse_to_test_output;
use std::fs;

mod util;

#[test]
fn mixed_content() {
    insta::glob!("input/mixed/mixed_content_*", |path| {
        let post = fs::read_to_string(path).unwrap();
        insta::assert_debug_snapshot!(parse_to_test_output(&post));
    });
}
