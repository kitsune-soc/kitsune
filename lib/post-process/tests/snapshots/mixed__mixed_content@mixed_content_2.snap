---
source: lib/post-process/tests/mixed.rs
expression: parse_to_test_output(&post)
input_file: lib/post-process/tests/input/mixed/mixed_content_2
---
[
    (
        Link(
            "https://github.com/aumetra",
        ),
        37..63,
        "https://github.com/aumetra",
    ),
]
